File: pegasusX/packages/mobile-android-design/src/main/java/com/pegasus/design/MarketPack.kt
package com.pegasus.design

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.atomic.AtomicReference

data class MarketPack(
    val code: String,
    val name: String,
    val timezone: String,
    val currencyCode: String,
    val fiscalAdapter: String,
    val mapsAdapter: String,
    val mapCenterLat: Double = 0.0,
    val mapCenterLng: Double = 0.0,
    val checkoutReadsThis: Boolean,
) {
    val receiptLabel: String get() = fiscalReceiptLabel(fiscalAdapter)
}

data class AuthSession(
    val marketCode: String,
    val homeCell: String,
    val apiUrl: String,
    val pack: MarketPack?,
    val checkoutReadsThis: Boolean,
)

fun fiscalReceiptLabel(adapter: String?): String {
    return when (adapter.orEmpty().trim().uppercase()) {
        "MY_SOLIQ" -> "Soliq"
        "COMMERCIAL", "PEGASUS", "FAKE" -> "commercial"
        "PEPPOL" -> "PEPPOL"
        "PLANNED", "" -> "planned"
        else -> adapter.orEmpty().trim().uppercase()
    }
}

fun packCurrency(pack: MarketPack?, fallback: String = ""): String {
    val code = pack?.currencyCode?.trim().orEmpty()
    return if (code.isNotEmpty()) code.uppercase() else fallback
}

/** Empty pack prints the number only. Never invents UZS. */
fun formatPackMoney(minor: Long, pack: MarketPack?, decimalPlaces: Int = 2): String {
    val places = decimalPlaces.coerceAtLeast(0)
    val denom = Math.pow(10.0, places.toDouble())
    val units = minor / denom
    val nf = java.text.NumberFormat.getNumberInstance(java.util.Locale.US)
    nf.minimumFractionDigits = 0
    nf.maximumFractionDigits = places
    val formatted = nf.format(units).replace(',', ' ')
    val ccy = packCurrency(pack)
    return if (ccy.isEmpty()) formatted else "$formatted $ccy"
}

fun sessionPackCurrency(): String = packCurrency(MarketPackStore.pack)

/** Stored/event currency, else session pack. Empty pack does not invent UZS. */
fun moneyCurrency(raw: String?): String {
    val fromEvent = raw?.trim().orEmpty()
    if (fromEvent.isNotEmpty()) return fromEvent.uppercase()
    return sessionPackCurrency()
}

data class PackMapCenter(val lat: Double, val lng: Double)

/** Shipped pack camera. Empty/planned pack does not invent Tashkent. */
fun packMapCenter(pack: MarketPack?): PackMapCenter? {
    val lat = pack?.mapCenterLat ?: 0.0
    val lng = pack?.mapCenterLng ?: 0.0
    if (lat == 0.0 && lng == 0.0) return null
    return PackMapCenter(lat, lng)
}

fun sessionMapCenter(): PackMapCenter? = packMapCenter(MarketPackStore.pack)

object MarketPackStore {
    private val ref = AtomicReference<AuthSession?>(null)

    val current: AuthSession? get() = ref.get()
    val pack: MarketPack? get() = current?.pack
    val sessionApiUrl: String? get() = current?.apiUrl?.takeIf { it.isNotBlank() }

    fun set(session: AuthSession?) {
        ref.set(session)
    }

    fun clear() {
        ref.set(null)
    }
}

object MarketPackBinder {
    fun fetch(baseUrl: String, token: String): AuthSession? {
        if (token.isBlank()) return null
        val pinned = CellApi.pinApiBaseUrl(baseUrl, CellApi.homeCellFromJwt(token), MarketPackStore.sessionApiUrl)
        val url = URL("${pinned.trimEnd('/')}/v1/auth/session")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            setRequestProperty("Authorization", "Bearer $token")
            connectTimeout = 15_000
            readTimeout = 15_000
        }
        return try {
            if (conn.responseCode !in 200..299) return null
            val body = conn.inputStream.bufferedReader().use { it.readText() }
            parseAuthSession(body)?.also { MarketPackStore.set(it) }
        } catch (_: Exception) {
            null
        } finally {
            conn.disconnect()
        }
    }

    fun parseAuthSession(json: String): AuthSession? {
        return try {
            val root = JSONObject(json)
            val packObj = root.optJSONObject("pack")
            val pack = packObj?.let {
                MarketPack(
                    code = it.optString("code"),
                    name = it.optString("name"),
                    timezone = it.optString("timezone"),
                    currencyCode = it.optString("currency_code"),
                    fiscalAdapter = it.optString("fiscal_adapter"),
                    mapsAdapter = it.optString("maps_adapter"),
                    mapCenterLat = it.optDouble("map_center_lat", 0.0),
                    mapCenterLng = it.optDouble("map_center_lng", 0.0),
                    checkoutReadsThis = it.optBoolean("checkout_reads_this", false),
                )
            }
            AuthSession(
                marketCode = root.optString("market_code"),
                homeCell = root.optString("home_cell"),
                apiUrl = root.optString("api_url"),
                pack = pack,
                checkoutReadsThis = root.optBoolean("checkout_reads_this", false),
            )
        } catch (_: Exception) {
            null
        }
    }
}

File: pegasusX/packages/mobile-ios-core/Sources/PegasusNetworking/MarketPack.swift
import Foundation
import SwiftUI

struct MarketPack: Sendable, Equatable {
    var code: String
    var name: String
    var timezone: String
    var currencyCode: String
    var fiscalAdapter: String
    var mapsAdapter: String
    var mapCenterLat: Double
    var mapCenterLng: Double
    var checkoutReadsThis: Bool

    var receiptLabel: String { fiscalReceiptLabel(fiscalAdapter) }
}

struct AuthSession: Sendable, Equatable {
    var marketCode: String
    var homeCell: String
    var apiUrl: String
    var pack: MarketPack?
    var checkoutReadsThis: Bool
}

func fiscalReceiptLabel(_ adapter: String?) -> String {
    switch (adapter ?? "").trimmingCharacters(in: .whitespacesAndNewlines).uppercased() {
    case "MY_SOLIQ": return "Soliq"
    case "COMMERCIAL", "PEGASUS", "FAKE": return "commercial"
    case "PEPPOL": return "PEPPOL"
    case "PLANNED", "": return "planned"
    default: return (adapter ?? "").trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
    }
}

func packCurrency(_ pack: MarketPack?, fallback: String = "") -> String {
    let code = pack?.currencyCode.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    return code.isEmpty ? fallback : code.uppercased()
}

/// Empty pack prints the number only. Never invents UZS.
func formatPackMoney(_ minor: Int64, pack: MarketPack?, decimalPlaces: Int = 2) -> String {
    let denom = pow(10.0, Double(max(decimalPlaces, 0)))
    let units = Double(minor) / denom
    let formatted = units.formatted(.number.precision(.fractionLength(0...max(decimalPlaces, 0))))
        .replacingOccurrences(of: ",", with: " ")
    let ccy = packCurrency(pack)
    return ccy.isEmpty ? formatted : "\(formatted) \(ccy)"
}

/// Stored/event currency, else session pack. Empty pack does not invent UZS.
func displayPackCurrency(_ raw: String?, pack: MarketPack? = MarketPackStore.pack) -> String {
    let fromEvent = (raw ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
    if !fromEvent.isEmpty { return fromEvent.uppercased() }
    return packCurrency(pack)
}

/// Shipped pack camera. Empty/planned pack does not invent Tashkent.
func packMapCenter(_ pack: MarketPack?) -> (lat: Double, lng: Double)? {
    guard let pack else { return nil }
    if pack.mapCenterLat == 0 && pack.mapCenterLng == 0 { return nil }
    return (pack.mapCenterLat, pack.mapCenterLng)
}

func packMapCoordinate(_ pack: MarketPack? = MarketPackStore.pack) -> (lat: Double, lng: Double) {
    packMapCenter(pack) ?? (0, 0)
}

func pinnedAPIBaseURL(bootstrap: URL) -> URL {
    let raw = CellApi.pinApiBaseUrl(
        bootstrap: bootstrap.absoluteString,
        homeCell: CellApi.homeCellFromJwt(CellTokenCache.token),
        sessionApiUrl: MarketPackStore.sessionApiUrl
    )
    let withSlash = raw.hasSuffix("/") ? raw : raw + "/"
    return URL(string: withSlash) ?? bootstrap
}

enum CellTokenCache {
    private static let lock = NSLock()
    private static var value = ""

    static var token: String {
        get { lock.lock(); defer { lock.unlock() }; return value }
        set { lock.lock(); value = newValue; lock.unlock() }
    }
}

enum MarketPackStore {
    private static let lock = NSLock()
    private static var session: AuthSession?

    static var current: AuthSession? {
        lock.lock(); defer { lock.unlock() }
        return session
    }

    static var pack: MarketPack? { current?.pack }

    static var sessionApiUrl: String? {
        let url = current?.apiUrl.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return url.isEmpty ? nil : url
    }

    static func set(_ next: AuthSession?) {
        lock.lock(); session = next; lock.unlock()
    }

    static func clear() {
        set(nil)
    }
}

enum MarketPackBinder {
    static func fetch(baseUrl: String, token: String) async -> AuthSession? {
        guard !token.isEmpty else { return nil }
        let pinned = CellApi.pinApiBaseUrl(
            bootstrap: baseUrl,
            homeCell: CellApi.homeCellFromJwt(token),
            sessionApiUrl: MarketPackStore.sessionApiUrl
        )
        guard let url = URL(string: "\(pinned)/v1/auth/session") else { return nil }
        var req = URLRequest(url: url)
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        guard let (data, resp) = try? await URLSession.shared.data(for: req),
              let http = resp as? HTTPURLResponse,
              (200...299).contains(http.statusCode),
              let parsed = parse(data)
        else { return nil }
        MarketPackStore.set(parsed)
        return parsed
    }

    static func parse(_ data: Data) -> AuthSession? {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        var pack: MarketPack?
        if let obj = root["pack"] as? [String: Any] {
            pack = MarketPack(
                code: obj["code"] as? String ?? "",
                name: obj["name"] as? String ?? "",
                timezone: obj["timezone"] as? String ?? "",
                currencyCode: obj["currency_code"] as? String ?? "",
                fiscalAdapter: obj["fiscal_adapter"] as? String ?? "",
                mapsAdapter: obj["maps_adapter"] as? String ?? "",
                mapCenterLat: (obj["map_center_lat"] as? NSNumber)?.doubleValue ?? 0,
                mapCenterLng: (obj["map_center_lng"] as? NSNumber)?.doubleValue ?? 0,
                checkoutReadsThis: obj["checkout_reads_this"] as? Bool ?? false
            )
        }
        return AuthSession(
            marketCode: root["market_code"] as? String ?? "",
            homeCell: root["home_cell"] as? String ?? "",
            apiUrl: root["api_url"] as? String ?? "",
            pack: pack,
            checkoutReadsThis: root["checkout_reads_this"] as? Bool ?? false
        )
    }
}

/// GS-R splash: currency + receipts.
struct PackBanner: View {
    let pack: MarketPack?

    var body: some View {
        if let pack, !pack.currencyCode.isEmpty {
            Text("\(pack.currencyCode) · receipts: \(pack.receiptLabel)")
                .font(.caption)
                .padding(.horizontal, 10)
                .padding(.vertical, 4)
                .overlay(Capsule().strokeBorder(.secondary, lineWidth: 1))
                .accessibilityIdentifier("gs-r-pack-chip")
        }
    }
}
